### PR TITLE
US-1129 Fix do not override built history

### PR DESCRIPTION
**Pull request by laurgarn**

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 1]
```diff
@@ -629,29 +706,52 @@ public void notifyModificationNodeChanged(UUID studyUuid, UUID nodeUuid) {
...
+    private void fillNodesInBuildOrder(NodeEntity nodeEntity, boolean nodeOnlyReport,
+        Map<UUID, Pair<UUID, String>> defNodeIdToReport,
+        List<Pair<UUID, String>> uuidsAndNames) {
+
         AbstractNode node = repositories.get(nodeEntity.getType()).getNode(nodeEntity.getIdNode());
-        res.add(0, Pair.of(getReportUuid(nodeEntity.getIdNode()), node.getName()));
-        if (node.getType() == NodeType.NETWORK_MODIFICATION && !nodeOnlyReport) {
-            getParentReportUuidsAndNamesFromNode(nodeEntity.getParentNode(), false, res);
+        if (nodeEntity.getType() != NodeType.NETWORK_MODIFICATION) {
+            uuidsAndNames.add(0, Pair.of(node.getReportUuid(), "Root"));
```
Comment (niconoir) [inline src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:715]: Better use ROOT_NODE_NAME instead of "Root"

**Comment by dbraquart** [comment]
The US nb mentioned in the PR title is wrong. In any case, I believe we should not mention it here.

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 2]
```diff
@@ -544,11 +581,51 @@ public void afterCommit() {
...
+                .map(ReportUsageEntity::getId)
+                .collect(Collectors.toList());
+            reportsUsagesRepository.deleteAllByIdInBatch(ownUsagesUuids);
+
+            UUID reportUuid = modificationNode.getReportUuid();
+            String variantId = modificationNode.getVariantId();
+            ownUsedReportIds.add(reportUuid);
+
+            invalidateNodeInfos.addVariantId(variantId);
+            ownUsedReportIds.removeAll(otherUsedReportIds);
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:612]: Yes, when a common ancestor is already built at the time some descendant on different branches are built, the reports of its building are part of the building of these descendant. So, when one of the descendant gets invalidate, we don't want to lose the shared reports. We must delete reports only when last usage is revoked.

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 3]
```diff
@@ -629,29 +706,52 @@ public void notifyModificationNodeChanged(UUID studyUuid, UUID nodeUuid) {
...
+                    && (!nodeOnlyReport || us.getDefinitionNode().getIdNode().equals(nodeUuid)))  {
+                    NodeEntity definitionNodeEntity = us.getDefinitionNode();
+                    AbstractNode definitionNode = repositories.get(definitionNodeEntity.getType()).getNode(definitionNodeEntity.getIdNode());
+                    defNodeIdToReport.put(definitionNodeEntity.getIdNode(),
+                        Pair.of(us.getReportId(), definitionNode.getName()));
+                }
+            });
+
+            fillNodesInBuildOrder(buildNodeEntity, nodeOnlyReport, defNodeIdToReport, uuidsAndNames);
+            //AbstractNode buildNode = repositories.get(buildNodeEntity.getType()).getNode(buildNodeEntity.getIdNode());
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:744]: Removed

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java (hunk 1 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:715]: Done

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 4]
```diff
@@ -544,11 +581,51 @@ public void afterCommit() {
...
+                    ownUsedReportIds.add(u.getReportId());
+                } else {
+                    otherUsedReportIds.add(u.getReportId());
+                }
+            });
+            List<UUID> ownUsagesUuids = usages.stream()
+                .filter(u -> u.getBuildNode().getIdNode().equals(node.getIdNode()))
+                .map(ReportUsageEntity::getId)
+                .collect(Collectors.toList());
+            reportsUsagesRepository.deleteAllByIdInBatch(ownUsagesUuids);
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:605]: Should I go with a 'reportUsageUuis' in the InvalidateNodeInfos ?

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 5]
```diff
@@ -544,11 +581,51 @@ public void afterCommit() {
         });
     }
 
-    private void invalidateChildrenBuildStatus(NodeEntity nodeEntity, List<UUID> changedNodes, boolean invalidateOnlyChildrenBuildStatus, InvalidateNodeInfos invalidateNodeInfos) {
+    private void fillInvalidateNodeInfos(NodeEntity node, InvalidateNodeInfos invalidateNodeInfos, boolean removeOnlyResults) {
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:584]: Done

File: src/main/java/org/gridsuite/study/server/dto/DeleteStudyInfos.java [hunk 6]
```diff
@@ -25,4 +25,6 @@ public class DeleteStudyInfos {
     private UUID networkUuid;
 
     private List<NodeModificationInfos> nodesModificationInfos;
+
+    private List<UUID> buildReportsUuids;
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/dto/DeleteStudyInfos.java:29]: Yes it should ! Thanks for spotting that !

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 7]
```diff
@@ -469,34 +503,49 @@ public LoadFlowInfos getLoadFlowInfos(UUID nodeUuid) {
...
         if (node.getType() == NodeType.NETWORK_MODIFICATION) {
             NetworkModificationNode modificationNode = (NetworkModificationNode) node;
             if (modificationNode.getBuildStatus() != BuildStatus.BUILT) {
-                buildInfos.insertModificationGroupAndReport(modificationNode.getModificationGroupUuid(), getReportUuid(nodeEntity.getIdNode()));
+                UUID reportUuid;
+                if (nodeEntity.getIdNode().equals(toBuildNode.getIdNode())) {
+                    reportUuid = modificationNode.getReportUuid();
+                } else {
+                    reportUuid = UUID.randomUUID();
+                    reportsUsagesRepository.save(new ReportUsageEntity(null, reportUuid, toBuildNode, nodeEntity));
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:516]: Would 'prepareBuild' be a better name than fillBuildInfos ?

File: src/main/java/org/gridsuite/study/server/dto/GroupAndReportInfos.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package org.gridsuite.study.server.dto;
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/dto/GroupAndReportInfos.java:1]: Added

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 9]
```diff
@@ -60,6 +65,8 @@
 @Service
 public class NetworkModificationTreeService {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(NetworkModificationTreeService.class);
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:68]: Removed

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/repositories/ReportUsageRepository.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package org.gridsuite.study.server.networkmodificationtree.repositories;
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/repositories/ReportUsageRepository.java:1]: Added

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/entities/ReportUsageEntity.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
+package org.gridsuite.study.server.networkmodificationtree.entities;
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/entities/ReportUsageEntity.java:1]: Added

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (niconoir) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:611]: Ok, thanks !

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (niconoir) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:514]: Yes, I would prefer.

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (niconoir) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:604]: Yes, and remove all reportUsages at the end of the NetworkModificationTreeService.invalidateBuild method, which is transactional

[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java]: Done

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:514]: Done

**Pull request merged**